lib/TPP/DefaultTppPasses.cpp

```diff
@@ -97,25 +100,48 @@ struct DefaultTppPasses : public DefaultTppPassesBase<DefaultTppPasses> {
     pm.addPass(bufferization::createDropEquivalentBufferResultsPass());
     pm.addNestedPass<func::FuncOp>(
         bufferization::createFinalizingBufferizePass());
-    // Clean up after bufferization
+    // Clean up after bufferization.
+    pm.addPass(bufferization::createBufferDeallocationPass());
     pm.addNestedPass<func::FuncOp>(createCanonicalizerPass());
 
-    // Convert all higher level dialects to TPP
+    // Convert generics to BRGEMM.
+    // The mapping is done after bufferization as the buffer semantics
+    // allow direct use of scf.parallel loops. This prevents different
+    // lowering outputs between input linalg on tensors and memrefs.
+    pm.addNestedPass<func::FuncOp>(createMapToBatchReduceGEMMPass());
+
+    // Convert all higher level dialects to TPP.
     pm.addNestedPass<func::FuncOp>(createConvertLinalgToTppPass());
     pm.addPass(createConvertVNNIToTppPass());
 
-    // Lower all TPP ops
+    // Lower all TPP ops.
     if (tppToLoops)
       pm.addNestedPass<func::FuncOp>(createConvertTppToLoopsPass());
     else
       pm.addNestedPass<func::FuncOp>(createConvertTppToXsmmPass());
-    // Lower all XSMM ops
-    pm.addPass(createConvertXsmmToFuncPass());
-    // Lower all Check ops
+
+    // Lower all Check ops.
     pm.addPass(createConvertCheckToLoopsPass());
-    // Lower all LinalgX ops
+    // Lower all LinalgX ops.
     pm.addPass(createLinalgXToLoopsPass());
 
+    // Postprocess generated loops.
+    // Perform LICM before function calls are generated to ensure that ops which
+    // map directly to functions also get moved outside of loops, if possible.
+    // This approach assumes that the function calls do not have any side
```

Comment:
maybe specify that LICM is used for the dispatch

Reply:
While this was the original motivation, it is still a generic pass that will move anything it can.
So, I'll leave it out as the exact ops it affects are not (or at least shouldn't be) relevant.